lib/cchain/provider/abci.go

```diff
@@ -763,3 +671,134 @@ func attsFromAtHeight(ctx context.Context, cl atypes.QueryClient, chainVer xchai
 
 	return atts, true, nil
 }
+
+// binarySearch uses a binary search between defined start and end consensus block heights to
+// find the attestations with the provided fromOffset.
+func binarySearch(
+	ctx context.Context,
+	attCl atypes.QueryClient,
+	chainVer xchain.ChainVersion,
+	chainName string,
+	fromOffset uint64,
+	startHeight uint64,
+	endHeight uint64,
+) (uint64, error) {
+	const endpoint = "search_offset"
+
+	for steps := 0; startHeight <= endHeight; steps++ {
+		midHeight := startHeight + umath.SubtractOrZero(startHeight, startHeight)/2
+
+		earliestAtt, ok, err := queryEarliestAttestation(ctx, attCl, chainVer, midHeight)
+		if err != nil {
+			incQueryErr(endpoint)
+			return 0, errors.Wrap(err, "abci query earliest-attestation-in-state")
+		}
+		if !ok {
```

Comment:
nit: I like to combine all validation into a single logical block, so prefer to use `else if`